### PR TITLE
feat: add SpaceCleanup controller

### PR DESCRIPTION
**Pull request by MatousJobanek**
add SpaceCleanup controller that deletes every space that:
* doesn't have any SpaceBinding
* it was created more than 30 seconds ago

paired with https://github.com/codeready-toolchain/toolchain-e2e/pull/479

### CODE REVIEW

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+		require.NoError(t, err)
+		assert.True(t, res.Requeue)
+		assert.LessOrEqual(t, res.RequeueAfter, time.Second)
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, space.Name, cl).
+			Exists()
+	})
+
+	t.Run("with SpaceBinding - Space shouldn't be deleted", func(t *testing.T) {
+		// given
+		space := spacetest.NewSpace("without-spacebinding", spacetest.WithCreationTimestamp(time.Now().Add(-time.Minute)))
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:60]:
```suggestion
		space := spacetest.NewSpace("with-spacebinding", spacetest.WithCreationTimestamp(time.Now().Add(-time.Minute)))
```

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+	t.Run("when is already being deleted", func(t *testing.T) {
+		// given
+		space := spacetest.NewSpace("being-deleted", spacetest.WithDeletionTimestamp())
+		r, req, cl := prepareReconcile(t, space)
+
+		// when
+		_, err := r.Reconcile(context.TODO(), req)
+
+		// then
+		require.NoError(t, err)
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:83]:
```suggestion
		require.NoError(t, err)
		assert.False(t, res.Requeue)
```

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+		spacetest.AssertThatSpace(t, test.HostOperatorNs, space.Name, cl).
+			Exists()
+	})
+
+	t.Run("when the space is not there, then just skip it", func(t *testing.T) {
+		// given
+		space := spacetest.NewSpace("not-found")
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:90]:
```suggestion
		space := spacetest.NewSpace("not-found") // will be disregarded, only included for call to prepareReconcile func
```

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+		r, req, _ := prepareReconcile(t, space)
+		empty := test.NewFakeClient(t)
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:92]:
```suggestion
		empty := test.NewFakeClient(t) // with space disregarded
```

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+		empty.MockList = func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+			return fmt.Errorf("shouldn't be called")
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:94]: Nice, didn't think about that 🙂

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,184 @@
...
+			r, req, cl := prepareReconcile(t, space)
+			cl.MockGet = func(ctx context.Context, key client.ObjectKey, obj client.Object) error {
+				return fmt.Errorf("some error")
+			}
+
+			// when
+			res, err := r.Reconcile(context.TODO(), req)
+
+			// then
+			require.Error(t, err)
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:120]: Should we check the error message here to be sure it's the right one?

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 7]
```diff
@@ -0,0 +1,184 @@
...
+			r, req, cl := prepareReconcile(t, space)
+			cl.MockList = func(ctx context.Context, list client.ObjectList, opts ...client.ListOption) error {
+				return fmt.Errorf("some error")
+			}
+
+			// when
+			res, err := r.Reconcile(context.TODO(), req)
+
+			// then
+			require.Error(t, err)
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:139]: Here too

File: controllers/spacecleanup/space_cleanup_controller_test.go [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+			r, req, cl := prepareReconcile(t, space)
+			cl.MockDelete = func(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+				return fmt.Errorf("some error")
+			}
+
+			// when
+			res, err := r.Reconcile(context.TODO(), req)
+
+			// then
+			require.Error(t, err)
```
Comment (rajivnathan) [inline controllers/spacecleanup/space_cleanup_controller_test.go:157]: also here

**Review by rajivnathan — APPROVED** [review]
Nice and clean! 🧹 Just added a few very minor comments/suggestions

File: controllers/spacecleanup/space_cleanup_controller_test.go (hunk 7 above)
Comment (MatousJobanek) [reply controllers/spacecleanup/space_cleanup_controller_test.go]: done in 667aa48c

File: controllers/spacecleanup/space_cleanup_controller_test.go (hunk 8 above)
Comment (MatousJobanek) [reply controllers/spacecleanup/space_cleanup_controller_test.go]: done in 667aa48c

**Comment by MatousJobanek** [comment]
@rajivnathan thanks a lot for your review. :pray:  I addressed all your comments :+1:

**Review by rajivnathan — APPROVED** [review]

**Pull request merged**